packages/server/postgres/migrations/1613005669863_template-refs.ts

```diff
@@ -0,0 +1,25 @@
+import {ColumnDefinitions, MigrationBuilder} from 'node-pg-migrate'
+
+export const shorthands: ColumnDefinitions | undefined = undefined
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+  pgm.sql(`
+    CREATE TABLE IF NOT EXISTS "TemplateScaleRef" (
+      "id" CHAR(24) PRIMARY KEY,
+      "scale" JSONB NOT NULL,
+      "createdAt" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
+    );
+    CREATE TABLE IF NOT EXISTS "TemplateRef" (
+      "id" CHAR(24) PRIMARY KEY,
+      "template" JSONB NOT NULL,
```

Comment:
i'm glad you brought this up!
i included `name` in the JSON because `id = md5(template)` if i take the name out of it, then to perform a checksum you'd have to say `id = md5({dimensions, name})`. My thought is that it's more intuitive to have a single param.

Does knowing this change your opinion on it at all? Either is possible, just not sure which is more clear

Reply:
I guess I still have the tension here. The single column with JSONB type can easily outgrow fast, say we introduce new field to template, e.g., "description", "emoji", "color", etc.

My understanding for hashing the entire content to generate the id is when you insert into Postegres, you do `ON CONFLICT (id) DO NOTHING ` to avoid duplicates. So, with that, why couldn't you just do `ON CONFLICT (name, dimensions) DO NOTHING`?

Reply:
`ON CONFLICT (name, dimensions)` is O(n) instead of O(log<sub>t</sub>n) for an ID which has a btree index (in practice, it's closer to O(1) with minimization factor t).

Reply:
OK you convinced me! Also I checked the code again and I realized one probably wouldn't deal with raw PG column data in the codebase directly. Rather to query `TemplateRef` table, you will call `getTemplateRefById` method which has strong type for the returned result:

```
interface TemplateRef extends IGetTemplateRefByIdQueryResult {
  dimensions: {
    name: string
    scaleRefId: string
  }[]
}
```

This solves my initial tension on "how the hell would I know what attributes are there in TemplateRef"